Multiple file uploads support

dmitry-viskov opened the pull request:
Continuation of the work that was started here: https://github.com/edx/edx-ora2/pull/944

Ability to attach more than one file:

![ora_multi](https://cloud.githubusercontent.com/assets/1876893/21913860/554293aa-d940-11e6-8e41-a65700720ba5.jpeg)

@cahrens @marcotuts please take a look

edx-webhook commented:
Thanks for the pull request, @dmitry-viskov! I've created [OSPR-1646](https://openedx.atlassian.net/browse/OSPR-1646) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

dmitry-viskov commented:
@cahrens @marcotuts could somebody review this PR please?

cahrens commented:
@gsong it doesn't look like this PR is linked to an OSPR.

gsong commented:
@cahrens It's <https://openedx.atlassian.net/browse/OSPR-1646>.

shellyu commented:
Continuing the conversation from [944](https://github.com/edx/edx-ora2/pull/944#issuecomment-263308025):

I'm concerned with the cumulative 5mb limit because several MIT courses already find that to be a challenge with a single file upload (21W.789x, SysEngX). Also, would we expect the learners to calculate total file size in addition to uploading projects on a deadline? I worry confusion about that would increase the already high course team support required for using ORA problems (and possibly learner support @ edX too.)

baldwin47 commented:
@cahrens @marcotuts thoughts on the file size concern expressed by @shellyu ? We're hoping to begin work on https://github.com/edx/edx-ora2/pull/970 soon. It would be good to take care of both of these issues at once

marcotuts commented:
Hi everyone, here is a quick summary / next step suggestion:

In the 944 pull request I mentioned the following:
"In the future we could have a separate story introduce a second file size max cap that is higher than 5mb. That way for example you could set MAX-FILE-SIZE to 5mb but a new MAX-UPLOAD-TOTAL-SIZE (or something with a better name than what I suggested :) to be 25mb. That way learners could upload 5 different ~5mb sized files, which limits the accidental upload of many more files."

For this PR I would make the following suggestion:
*Continue to limit max file size cumulative to 5MB** This wouldn't fully address Shelly's concern, at least not with this PR. Separately, I will connect with edX folks on the file size limit to determine our rationale around that size. 

As an aside, @shellyu  - what size files are those 2 courses currently attempting to attach? Is this resolvable by having a slightly higher file size or are the files they're trying to upload best left as a link / url in the prompt (if they're much larger than 5MB for example)? 

In a separate PR we could look at using both the per-file max (5mb) and a new cumulative max file size configuration (ex: 20MB) in the future.

shellyu commented:
Thanks for the update @marcotuts! 

One team was looking for 10MB and another was looking for more, but I can't locate the actual request anymore... For either, even a small increase would help.

Both are project-based courses and quality responses might require substantial use of images, diagrams, etc. One team has tried external links submitted in the text area and the related support effort is high (staff grade overrides, asking students to make files public, managing the discussion posts related to all this...) After several runs using both approaches, the team still prefers the simplicity of an in-problem upload that has a limit large enough to handle image-heavy PDFs.

dmitry-viskov commented:
@cahrens @gsong @marcotuts could somebody review this PR please?

sstack22 commented:
@dmitry-viskov - thanks so much! We'll be looking at this in our next sprint. I'll also confirm with our team if we have any firm requirements on file size limiy. I think it ought to be fine to increase the limit to 10 MB, but I'll validate that. (If it's easier to make the adjustment in a separate PR, that makes sense).

sstack22 commented:
Hi @dmitry-viskov - this is in our current sprint (which just started). We had been prioritizing finishing out the earlier ORA OSPRs, which is why this was not addressed sooner. Hope this clarifies!

efischer19 commented:
@jibsheet and I are looking to see if it would be okay to bump `MAX_FILES_SIZE` up to 10MB, we'll have an answer for that before this feature gets released.

@sstack22 (and anyone else), I've set up this feature on my sandbox for testing if you're curious. https://efischer19.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/courseware/c07925b4b38c443f82b4f7dc27a68dfb/be1746b7c5c54ff99b783e6c5e8bae7e/1

dmitry-viskov commented:
hi @efischer19 . Thanks for the review

>Given that, I think the path forward will be to merge each of the 3 PRs (this, #993, and one yet to be opened correct?) into master separately. 

ok. I fine with that. For now I'm busy with the other activity (that doesn't connected with ORA) and will have a time to start working on the last PR. May be in 1 or 2 weeks from this moment i'll have some time to begin working on that. The problem is that I don't want to create one big PR with all three features because it will be very big and complicated. If you want I may recreate this PR and choose other branch for merge (I mean not `master`)

>my biggest concern is that I can't decrease the number of files attached to my response after attaching some number initially

I may suggest to remove all uploaded files on every attempt to upload files.  What do you think?

>I'm also a bit concerned about acceptance test coverage - would you mind modifying FileUploadTest.test_file_upload in acceptance.py to upload 2 files

Unfortunately i did not found the way how to upload many files through `selenium.webdriver.common.action_chains.ActionChains.send_keys` from the python selenium library. If you know how to do this - please let me know :slightly_smiling_face:

efischer19 commented:
> If you want I may recreate this PR and choose other branch for merge (I mean not master)

A feature branch is a great idea; I'll create one and point this PR at it (FYI- You can do this from the "edit" button at the top of a PR now, no recreation needed).

> I may suggest to remove all uploaded files on every attempt to upload files. What do you think?

Love it! Simple and elegant. 👍 from me.

> Unfortunately i did not found the way how to upload many files through selenium.webdriver.common.action_chains.ActionChains.send_keys from the python selenium library. If you know how to do this - please let me know 🙂

I'll take a quick poke and see what I can do here. I think that we're essentially sending a full file path across as the "mock result" of that dialog box, surely there's a way to do that twice.

efischer19 commented:
@dmitry-viskov I think I've gotten the acceptance tests updated here, can you cherry-pick? https://github.com/edx/edx-ora2/commit/e545957af5909e793024697d1b8eaa0d8f4e653a

srpearce commented on openassessment/templates/openassessmentblock/oa_uploaded_file.html:
```diff
@@ -10,18 +10,24 @@ <h5 class="{{ class_prefix }}__display__title">
+                     src="{{ file_url }}" />
+            {% elif file_upload_type == "pdf-and-image" or file_upload_type == "custom" %}
+                <a href="{{ file_url }}" class="submission__answer__file submission--file" target="_blank">
+                    {% trans "View the file associated with this submission:" %} #{{ forloop.counter }}
```
View the files...

srpearce commented on openassessment/templates/openassessmentblock/oa_uploaded_file.html:
```diff
@@ -10,18 +10,24 @@ <h5 class="{{ class_prefix }}__display__title">
+        {% endfor %}
+        </div>
+        {% if show_warning %}
+            <p class="submission_file_warning">{% trans "(Caution: These files were uploaded by another course learner and has not been verified, screened, approved, reviewed, or endorsed by the site administrator. If you decide to access it, you do so at your own risk.)" %}</p>
```
Caution: These files were uploaded by another course learner and have not been verified, screened, approved, reviewed, or endorsed by the site administrator. If you decide to access the files, you do so at your own risk.)

dmitry-viskov commented:
@efischer19 @srpearce 
hi. thanks for the comments. I've updated the PR.

@efischer19 I've implemented `BaseBackend.remove_file` function to remove keys from storage (S3 + local file system). Unfortunately I'm not very familiar with Swift. As I think implementation of this function is not very important because basic files upload functionality will work without it too (+ Swift is not so popular solution to store files). If you have any Swift bucket for testing I may try to correct this deficiency or let's postpone it for the future.

Also now after you changed branch for merge - I see that probably I was wrong. There is no travis tests and after I made rebase locally now there are a lot of code from the other commits. I may suggest to return branch back to master. During this and next week I'm going to finish development of all missing description fields for files and create the last PR. I may suggest to combine all these three PRs into one and create one big PR

efischer19 commented:
@dmitry-viskov I ran out of time to get to this today, I'll give it a closer look Monday and we'll keep moving. On first glance this looks great!

Sorry about the branch confusion - let's go ahead and return this PR to point at master. If I may suggest, it would be fairly simple to chain your PRs on top of each other instead of having one big one:

- first PR: branch 1 -> master
- second PR: branch 2 -> branch 1; update to branch 2 -> master after PR 1 merges
- third PR: branch 3 -> branch 2; update to branch 3 -> master after PR 2 merges

This will make it possible to work with/review code in more manageable chunks, yes?

dmitry-viskov commented:
hi @efischer19 

>first PR: branch 1 -> master
>second PR: branch 2 -> branch 1; update to branch 2 -> master after PR 1 merges
>third PR: branch 3 -> branch 2; update to branch 3 -> master after PR 2 merges

mmmm.... not quite sure that I understand you. Are you going to merge each branch to master step by step?
Now I am finalizing the development of files descriptions. So I may combine all three PRs into one - there is no problem for me. If you willn't have difficulty to check all this - i'm ok with this approach :-)

efischer19 commented:
@dmitry-viskov I've finally had a chance to look at this PR again - the code looks great, thanks for the updates. I'm satisfied, let's move on to the next PR.

> So I may combine all three PRs into one - there is no problem for me. If you willn't have difficulty to check all this - i'm ok with this approach :-)

Sounds great, whatever is easiest for you :) Are you going to add the new changes as commits on top of this branch? I'll start building a sandbox and taking a look at the code in https://github.com/edx/edx-ora2/pull/993 this afternoon (that's the next PR in this series, correct?), let me know where you want the "big" PR to be and I'll make comments there.

dmitry-viskov commented:
hi @efischer19 . Thanks for the review.
Yes. I'm going to update this PR (to add new commits)

dmitry-viskov commented:
@efischer19 Screenshots how files descriptions fields looks:

![001](https://cloud.githubusercontent.com/assets/1876893/24958442/f6d10420-1f97-11e7-8504-b99d08b30b8d.jpg)
![002](https://cloud.githubusercontent.com/assets/1876893/24958441/f6cceb06-1f97-11e7-9c1d-cb6aa1a3999e.jpg)
![003](https://cloud.githubusercontent.com/assets/1876893/24958440/f6cce6a6-1f97-11e7-9404-04e112ccd07f.jpg)
![004](https://cloud.githubusercontent.com/assets/1876893/24958443/f6d1d508-1f97-11e7-8895-abe33d19801e.jpg)

dmitry-viskov commented:
@efischer19 it is the second feature.
The third one i'm going to move today as the new commit from this PR: https://github.com/edx/edx-ora2/pull/993

dmitry-viskov commented:
![files_req](https://cloud.githubusercontent.com/assets/1876893/24983986/445855f8-1ff3-11e7-925e-1ba6e6aca8f0.jpeg)
![form_without_text_response](https://cloud.githubusercontent.com/assets/1876893/24983987/44590fde-1ff3-11e7-88c7-5da1cc0d687e.jpg)
![submitted_answer](https://cloud.githubusercontent.com/assets/1876893/24983988/4459fffc-1ff3-11e7-913d-7f323f56e002.jpeg)

dmitry-viskov commented:
@efischer19 I've just pushed the third (and the last) commit ^

efischer19 commented:
Excellent, thanks @dmitry-viskov. I'll take a look at this and get back to you later this afternoon.

efischer19 commented:
Hey @dmitry-viskov, sorry for the late reply. This is looking good so far, @sstack22 and I have taken an initial look and don't have any huge issues to raise. We're getting a few others to check out our sandbox internally, then we'll combine everyone's feedback at once instead of sending you a multitude of small requests.

efischer19 commented on openassessment/xblock/submission_mixin.py:
```diff
@@ -185,14 +192,76 @@ def save_submission(self, data, suffix=''):
+                        all_description_correct = False
+                        break
+
+                if all_description_correct:
```
This could be more succinct and eliminate the `all_description_correct` logic:

`if all(map(lambda description: isinstance(description, basestring), descriptions)):`

efischer19 commented on openassessment/templates/openassessmentblock/oa_uploaded_file.html:
```diff
@@ -10,18 +10,29 @@ <h5 class="{{ class_prefix }}__display__title">
+                {% if file_description %}
+                <div class="submission__file__description__label">{{ file_description }}:</div>
+                {% endif %}
+                <div><img class="submission__answer__file submission--image" src="{{ file_url }}" /></div>
```
See [this page](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-labelledby_attribute) for more details on using `aria-labelledby`, which is appropriate here due to the text already existing onscreen.

dmitry-viskov replied:
Updated

efischer19 commented on openassessment/xblock/static/js/src/lms/oa_response.js:
```diff
@@ -38,8 +41,8 @@ OpenAssessment.ResponseView.prototype = {
-    // Maximum file size (5 MB) for an attached file.
-    MAX_FILE_SIZE: 5242880,
+    // Maximum size (5 MB) for all attached files.
+    MAX_FILES_SIZE: 5242880,
```
Double it! Per @jibsheet, we're okay to raise this to 10485760 as part of this PR.

dmitry-viskov replied:
Done

dmitry-viskov commented:
hi @efischer19 . Thanks for review.

>Can we indicate to the user when the text of a response is required or optional?

Yes. But could you please provide mockups or screenshots how it should look like?

>We’ve made a slight change to how the description tagging should work, largely due to a desire to make thumbnails available during tagging. The desired plan of action is now:
>- Use the “Choose files” dialog to select files
>- Select “upload files” to upload them immediately
>- Expose the file__description textareas, with thumbnails display for images (this can be done in css, let me know if I can assist)
>- Save the descriptions on final submit (or error if they are left blank)

Not quite understand why the second step (I mean "Select upload files”) is required? May be I am wrong but do you suggest to upload files firstly and save descriptions only on the final "submit your response" step? If so - it will require a lot of changes in the existed functionality. If you want to make thumbnails available during tagging I may suggest to display images thumbnails immediately after user choose them (i.e after click "Choose fIles" button). It could be done using [HTML5 File API](https://developer.mozilla.org/en-US/docs/Using_files_from_web_applications). Example: http://html5demos.com/file-api

>It would also be nice to allow for removal of a single file during step #3 above - can you add a button to allow for this?

Does it really necessary to be done in the current PR? This button is absent in the current master implementation - so users can't remove their uploaded files. In the current PR users always can remove all their uploaded files by dint of uploading one new file (because every time on each upload we remove all previously uploaded files). I'm not complaining but this PR looks like endless improvement of one feature. it would be good to merge something - buttons to remove files could be added in the future PRs. It is not the main feature but also requires: 1. additional mockups/images, 2. additional development, 3. additional tests, 4. bugfix etc

efischer19 commented:
> Can we indicate to the user when the text of a response is required or optional?

> Yes. But could you please provide mockups or screenshots how it should look like?

Sure thing. I'll work with @sstack22 to get something to you.

> If you want to make thumbnails available during tagging I may suggest to display images thumbnails immediately after user choose them (i.e after click "Choose fIles" button). It could be done using HTML5 File API.

Oh man, this is awesome! Thanks for showing me! Thumbnail availability during tagging is exactly why we wanted to change things, this will be much simpler. We'll want to do some thorough testing to ensure browser compatibility ([looks like IE/Edge only partially support it](http://caniuse.com/#feat=fileapi)), I'd be glad to help out with that.

> It would also be nice to allow for removal of a single file during step #3 above - can you add a button to allow for this?

> Does it really necessary to be done in the current PR?

Fair point, we can add this in later as a separate PR.

dmitry-viskov commented:
@efischer19 [this example](https://jsfiddle.net/97b0e26e/) working fine in Edge, IE10 and IE11:

![fileapi_edge](https://cloud.githubusercontent.com/assets/1876893/25282554/858a807e-26b9-11e7-97f5-d8e9ff5cf449.jpeg)

![fileapi_ie10](https://cloud.githubusercontent.com/assets/1876893/25282559/8a3732f2-26b9-11e7-8766-9f89a7541af3.jpeg)

![fileapi_ie11](https://cloud.githubusercontent.com/assets/1876893/25282570/8f16f4d8-26b9-11e7-88d9-55a904e9e944.jpeg)

sstack22 commented:
@efischer19 , @dmitry-viskov  - the "Your response" line should just read "Your response (required)" or "Your response (optional)"

efischer19 commented:
Awesome, thanks @dmitry-viskov. That will work wonderfully, let's run with it.

dmitry-viskov commented on openassessment/fileupload/backends/swift.py:
```diff
@@ -63,6 +63,10 @@ def get_download_url(self, key):
+    def remove_file(self, key):
+        """
+        @TODO Need to be implemented to remove previously uploaded files
+        """
```
Mmm. Ok. But it this case all instances that uses swift storage will be unworkable. `remove` method is used in the new `remove_all_uploaded_files` `XBlock.json_handler`:
```
    @XBlock.json_handler
    def remove_all_uploaded_files(self, data, suffix=''):  # pylint: disable=unused-argument
        """
        Removes all uploaded user files.

        """
        removed_num = 0
        for i in range(self.MAX_FILES_COUNT):
            removed = file_upload_api.remove_file(self._get_student_item_key(i))
            if removed:
                removed_num += 1
            else:
                break
        return {'success': True, 'removed_num': removed_num}
```
But this API is called every time before each upload.

efischer19 replied:
Hmm, yeah that seems worse doesn't it? Let me investigate a bit here, maybe we can implement `remove_file` in swift without too much hassle.

dmitry-viskov replied:
@efischer19 Seems that I've found a solution. I've configured [virtualbox with swift](https://github.com/swiftstack/vagrant-swift-all-in-one) and this code could be used to remove files on swift storage:

```
temp_url = swiftclient.utils.generate_temp_url(
    path='%s/%s/%s' % (url.path, bucket_name, key_name),
    key=key,
    method='DELETE',
    seconds=3600)
remove_url = '%s://%s%s' % (url.scheme, url.netloc, temp_url)
response = requests.delete(remove_url)
```

dmitry-viskov replied:
Updated

dmitry-viskov commented on openassessment/templates/openassessmentblock/response/oa_response.html:
```diff
@@ -89,27 +89,31 @@ <h5 class="submission__answer__part__text__title">{% trans "The prompt for this
                                 </div>
+                                {% endif %}
                             </li>
                         {% endfor %}
```
Updated

dmitry-viskov commented on openassessment/templates/openassessmentblock/response/oa_response.html:
```diff
@@ -89,27 +89,31 @@ <h5 class="submission__answer__part__text__title">{% trans "The prompt for this
-                            <button type="submit" class="file__upload action action--upload" disabled>{% trans "Upload your file" %}</button>
+                            <input type="file" class="submission__answer__upload file--upload" id="submission_answer_upload_{{ xblock_id }}" multiple="">
+                            <div class="files__descriptions"></div>
+                            <button type="submit" class="file__upload action action--upload" disabled>{% trans "Upload files" %}</button>
```
Updated

dmitry-viskov commented:
@efischer19 Updated the PR. Does now it looks better?
![ora_update_new](https://cloud.githubusercontent.com/assets/1876893/25413494/7e20a3e0-2a33-11e7-8b60-f8cbce4e4c64.png)

efischer19 commented on openassessment/xblock/static/js/src/lms/oa_response.js:
```diff
@@ -474,71 +537,259 @@ OpenAssessment.ResponseView.prototype = {
+                });
+                img.onload = function() {
+                    window.URL.revokeObjectURL(this.src);
+                };
```
Can you give `img` the default alt attribute of `Thumbnail view of <file name>`? That will satisfy [accessibility requirements](https://dequeuniversity.com/rules/axe/2.1/image-alt)

dmitry-viskov replied:
fixed

efischer19 commented on openassessment/xblock/static/js/src/lms/oa_response.js:
```diff
@@ -474,71 +537,259 @@ OpenAssessment.ResponseView.prototype = {
+                descriptionsExists = false;
+            }
+            textarea.addClass('file__description file__description__' + i);
+            textarea.appendTo(divTextarea);
```
`textarea` needs [a label for accessibility](https://dequeuniversity.com/rules/axe/2.1/label)

dmitry-viskov replied:
fixed

efischer19 commented on openassessment/xblock/static/js/src/lms/oa_response.js:
```diff
@@ -474,71 +537,259 @@ OpenAssessment.ResponseView.prototype = {
         }
-        $(".file__upload").prop('disabled', this.files === null);
+
+        $(this.element).find('.file__upload').prop('disabled', !descriptionsExists);
```
How hard would it be to disable the `Submit your response` button while the descriptions are being edited? Currently it seems a bit strange to have it be available without doing anything. Could you disable it here, and re-enable in `removeFileDescriptions`?

<img width="860" alt="screen shot 2017-04-26 at 4 06 40 pm" src="https://cloud.githubusercontent.com/assets/7373924/25454480/92eb5ab0-2a9a-11e7-9997-11b006896ea7.png">

dmitry-viskov replied:
@efischer19 
There could be two different situations.

**1**. If in the ORA settings `Text Response == optional` and `File Uploads Response == required` in this case Submit button will be locked until user will upload at least one file.

**2**. If in the ORA settings `Text Response == optional` and also `File Uploads Response == optional` (I may suggest that your described case above is this one) - in this case Submit button will be unlocked if user will type something in the "response" textarea or if at least one file will be uploaded. File descriptions are required only for "Upload" button. So you may type anything in the textarea field, choose files, decide not to upload anything and submit the answer without upload.

Do you suggest to lock "Submit" button even if `File Uploads Response == optional` but files are chosen but not uploaded?

dmitry-viskov replied:
@efischer19 
![text_response_optional](https://cloud.githubusercontent.com/assets/1876893/25461510/1817e1b6-2af2-11e7-90d8-b8db5d392ee9.png)

efischer19 replied:
Sorry about the false alarm here! I had misconfigured my sandbox at some point, and [this warning](https://github.com/edx/edx-ora2/blob/89d31b652da5a63971fa2d7b0b80a176b5d45d8f/openassessment/xblock/static/js/src/lms/oa_response.js#L381) never popped up, that will handle the case I was concerned about.

<img width="912" alt="screen shot 2017-04-27 at 4 28 03 pm" src="https://cloud.githubusercontent.com/assets/7373924/25502935/859fd47c-2b66-11e7-8265-d90a9b6e76ed.png">

efischer19 commented:
Also, I forgot to call it out in the review I just submitted, but the new thumbnails-during-description-authoring is AWESOME, that looks much more friendly.

dmitry-viskov commented on openassessment/templates/openassessmentblock/oa_uploaded_file.html:
```diff
@@ -10,18 +10,30 @@ <h5 class="{{ class_prefix }}__display__title">
+            <div class="submission__answer__file__block submission__answer__file__block__{{ forloop.counter0 }}">
+            {% if file_upload_type == "image" %}
+                {% if file_description %}
+                <div class="submission__file__description__label" id="file_description_{{ xblock_id }}_{{ forloop.counter0 }}">{{ file_description }}:</div>
```
@efischer19 Good find. Thanks. I'll add `including_template` into templates. `xblock_id` is still necessary because it will be useful in case more than one ORA block on the same page

dmitry-viskov commented on openassessment/fileupload/backends/swift.py:
```diff
@@ -63,6 +63,24 @@ def get_download_url(self, key):
+            logger.exception(
+                u"An internal exception occurred while removing object on swift storage."
+            )
+            raise FileUploadInternalError(ex)
```
:+1:

dmitry-viskov commented:
@efischer19 thanks for your assistance! Commits are squashed.

marcotuts commented:
This is awesome! Thanks @dmitry-viskov and @efischer19 for your work on this. Huge improvements here. 💯
